### PR TITLE
feat: Automatically remind people to RSVP

### LINKED ISSUE #18: Automatically remind people to RSVP
_From @sman591 on February 1, 2017 20:0_

Just like we send follow-up emails for anyone that hasn't finished their application, we should send a follow-up email to those who haven't RSVP'd yet.

Loosely based on the following rules:
* Send follow-up 7 days after they were accepted (if signed up > 2 weeks before event)
* Send follow-up 5 days before event (if signed up > 10 days before event)
* Send follow-up 2 days before event (if signed up > 3 days before event)

_Copied from original issue: codeRIT/brickhack.io#472_

### PR DESCRIPTION
**Pull request by dquach217**
Fixes #18

### CODE REVIEW

**Comment by dquach217** [comment]
Currently Work in Progress, please do not approve as of yet

**Comment by dquach217** [comment]
@sman591 When you have the chance to, could you walk me through testing email triggers?

**Comment by sman591** [comment]
Happy to - I'll try to take a look later today or tomorrow and follow-up with some steps to test with

**Comment by dquach217** [comment]
@sman591 Make the changes as you requested.

File: app/models/user.rb [hunk 1]
```diff
@@ -38,6 +38,10 @@ def queue_reminder_email
     update_attribute(:reminder_sent_at, Time.now)
   end
 
+  def queue_rsvp_email
+    UserMailer.rsvp_reminder_email(id)
+  end
```
Comment (sman591) [inline app/models/user.rb]: This can be removed now that we're queueing on the questionnaire model

**Comment by dquach217** [comment]
@sman591 I think adding a test would be great, but not sure how we can replicate the actions since we would have it wait until it meets a certain criteria to be sent out

File: app/models/user.rb [hunk 2]
```diff
@@ -15,7 +15,7 @@ class User < ApplicationRecord
 
   validates_uniqueness_of :email
 
-  after_create :queue_reminder_email
+  after_create :queue_reminder_email, :queue_rsvp_email
```
Comment (sman591) [inline app/models/user.rb]:
```suggestion
  after_create :queue_reminder_email
```

**Review by sman591 — DISMISSED** [review]
For tests -- the best way to support the conditions would be to test against various `event_start_date` config values while wrapping the tests with `Timecop.freeze`

```ruby
HackathonConfig["event_start_date"] = "2020-06-15"
Timecop.freeze("2020-06-01") do
  # the tests here...
end
```

Up to you though

**Comment by dquach217** [comment]
I will try creating a bunch of tests using the suggested method. Thanks!

**Comment by cbaudouinjr** [comment]
Closing as PR is too stale.

**Comment by peterkos** [comment]
We'll release this as-is as part of 2.0 without the unit tests, as @cbaudouinjr has verified this functionality by hand. 
@dquach217 Is this ready to go otherwise?

**Review by cbaudouinjr — DISMISSED** [review]

**Review by peterkos — APPROVED** [review]

File: app/models/questionnaire.rb [hunk 3]
```diff
@@ -249,4 +250,18 @@ def queue_triggered_email_update
   def queue_triggered_email_create
     Message.queue_for_trigger("questionnaire.#{acc_status}", user_id)
   end
+
+  def queue_triggered_email_rsvp_reminder
+    if saved_change_to_acc_status? && acc_status == "accepted"
+      days_remaining = Date.parse(HackathonConfig["event_start_date"]) - Time.now.to_date
```
Comment (peterkos) [inline app/models/questionnaire.rb:256]: Time zone bug @cbaudouinjr told you ;)

**Review by peterkos — CHANGES_REQUESTED** [review]

File: app/mailers/user_mailer.rb [hunk 4]
```diff
@@ -20,4 +20,11 @@ def incomplete_reminder_email(user_id)
 
     Message.queue_for_trigger("user.24hr_incomplete_application", @user.id)
   end
+
+  def rsvp_reminder_email(user_id)
+    @user = User.find_by_id(user_id)
+    return if @user.blank? || !@user.questionnaire.acc_status == "accepted" || Time.now.to_date > Date.parse(HackathonConfig["event_start_date"])
```
Comment (peterkos) [inline app/mailers/user_mailer.rb:26]: Time zone bug 🤔

**Review by peterkos — CHANGES_REQUESTED** [review]

**Review by peterkos — APPROVED** [review]
LGTM! 🚀

**Pull request merged**